### PR TITLE
Add `build_collections` method to `CesmMetadataStoreCatalog`

### LINKED ISSUE #21: harmonize types
From the user perspective, our various classes are confusing. 

Should `manage_collections.CESMCollections` return a type of `CesmMetadataStoreCatalog`?

This would feel more natural. The return object from a freshly build collection could be query, etc. directly without having to re-invoke
```
collection = intake.open_cesm_metadatastore(collection_name)
```
for instance.

### LINKED ISSUE #27: Bug in `manage_collections.py`
A while back I changed `_extract_cesm_date_str` method to something I thought was  simple and correct. However, today when generating CESM collection catalog from files with `hsi`, I got this error:

```python-traceback
/gpfs/u/home/abanihi/devel/intake-cesm/intake_cesm/manage_collections.py in _build_cesm_collection_df_files(self, resource_key, resource_type, direct_access, filelist)
    233 
    234         for f in tqdm(filelist):
--> 235             fileparts = self._cesm_filename_parts(os.path.basename(f), self.component_streams)
    236 
    237             if fileparts is None:

/gpfs/u/home/abanihi/devel/intake-cesm/intake_cesm/manage_collections.py in _cesm_filename_parts(self, filename, component_streams)
    169 
    170         # define lists of stream strings
--> 171         datestr = self._extract_cesm_date_str(filename)
    172 
    173         for component, streams in component_streams.items():

/gpfs/u/home/abanihi/devel/intake-cesm/intake_cesm/manage_collections.py in _extract_cesm_date_str(self, filename)
    162     def _extract_cesm_date_str(self, filename):
    163         """Extract a datastr from file name."""
--> 164         b = filename.split(".")[-2]
    165         return b
    166 

IndexError: list index out of range
```

@matt-long, I was wondering whether you ever saw this error before? It appears that everything in `build_collections` notebook ran smoothly for you. I am not sure why It breaks on my part.

### PR DESCRIPTION
**Pull request by andersy005**
- Closes #21
- Fixes #27 
- Refactor tests:
    - Added sample NetCDF data
    - Allows testing collection build for posix `StorageResource` type
    - Add test for `to_xarray()` for collection entries with `direct_access=True`. See [example below](https://github.com/NCAR/intake-cesm/pull/28#issuecomment-461665600)

### CODE REVIEW

**Comment by andersy005** [comment]
@matt-long, This is ready for review. I tried to stick to the API you proposed in #21:

**Example**:
```python
In [1]: import intake                                                                                                                               

In [2]: from intake_cesm.config import set_options                                                                                                  

In [3]: set_options(database_directory="./tests/test_collections")                                                                                  
Out[3]: <intake_cesm.config.set_options at 0x118d31b38>

In [4]: build_args = {"collection_input_file" : "./tests/collection_input_test.yml"
    ...: , "collection_type_def_file" : "intake_cesm/cesm_definitions.yml"
   ...: , "overwrite_existing" : True}                                                                                                              

In [5]: col = intake.open_cesm_metadatastore(collection="cesm_dple_test_collection", build_args=build_args)                                         
INFO:root:Active collection : cesm_dple_test_collection
INFO:root:Active database: /Users/abanihi/devel/ncar/intake-cesm/tests/test_collections/cesm_dple_test_collection.csv
INFO:root:calling build
INFO:root:working on experiment: g.e11_LENS.GECOIAF.T62_g16.009
INFO:root:getting file listing: TEST-LIST:input-file:tests/intake-cesm-test-filelist
INFO:root:building file database: TEST-LIST:input-file:tests/intake-cesm-test-filelist
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████| 3/3 [00:00<00:00, 11234.74it/s]
1it [00:00, 93.86it/s]
INFO:root:working on experiment: g.e11_LENS.GECOIAF.T62_g16.009-copy
INFO:root:getting file listing: SAMPLE-DATA:posix:tests/sample_data
INFO:root:building file database: SAMPLE-DATA:posix:tests/sample_data
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████| 3/3 [00:00<00:00, 17873.45it/s]
1it [00:00, 57.33it/s]
Active collection: cesm_dple_test_collection

In [6]: col.df                                                                                                                                      
Out[6]: 
                                            resource resource_type direct_access  ... sequence_order has_ocean_bgc       grid
0                SAMPLE-DATA:posix:tests/sample_data         posix          True  ...              0           NaN  POP_gx1v6
1                SAMPLE-DATA:posix:tests/sample_data         posix          True  ...              0           NaN  POP_gx1v6
2                SAMPLE-DATA:posix:tests/sample_data         posix          True  ...              0           NaN  POP_gx1v6
3  TEST-LIST:input-file:tests/intake-cesm-test-fi...    input-file         False  ...              0           NaN  POP_gx1v6
4  TEST-LIST:input-file:tests/intake-cesm-test-fi...    input-file         False  ...              0           NaN  POP_gx1v6
5  TEST-LIST:input-file:tests/intake-cesm-test-fi...    input-file         False  ...              0           NaN  POP_gx1v6

[6 rows x 18 columns]

In [7]: cat = col.search(variable='O2', direct_access=True)                                                                                         

In [8]: ds = cat.to_xarray()                                                                                                                        

In [9]: ds                                                                                                                                          
Out[9]: 
<xarray.Dataset>
Dimensions:  (ens: 1, nlat: 5, nlon: 5, sigma: 5, time: 12)
Coordinates:
  * sigma    (sigma) float64 23.4 23.45 23.5 23.55 23.6
  * time     (time) float64 9.092e+04 9.094e+04 ... 9.122e+04 9.125e+04
Dimensions without coordinates: ens, nlat, nlon
Data variables:
    TLAT     (nlat, nlon) float64 dask.array<shape=(5, 5), chunksize=(5, 5)>
    ULONG    (nlat, nlon) float64 dask.array<shape=(5, 5), chunksize=(5, 5)>
    ULAT     (nlat, nlon) float64 dask.array<shape=(5, 5), chunksize=(5, 5)>
    TLONG    (nlat, nlon) float64 dask.array<shape=(5, 5), chunksize=(5, 5)>
    O2       (ens, time, sigma, nlat, nlon) float32 dask.array<shape=(1, 12, 5, 5, 5), chunksize=(1, 1, 5, 5, 5)>
```

File: intake_cesm/manage_collections.py [hunk 1]
```diff
@@ -364,8 +371,9 @@ def _build_cesm_collection(self, collection_attrs):
         # write data to csv
         self.df.to_csv(self.active_db, index=True)
 
-    def build_collections(self, overwrite_existing):
+    def _build_collections(self, overwrite_existing):
```
Comment (matt-long) [inline intake_cesm/manage_collections.py]: @andersy005, it strikes me that this routine is no-longer exactly appropriate in this class. It carries unnecessary logic to select the `collection_type`. I think that logic belongs in `CesmMetadataStoreCatalog.build_collection`. Let's imagine our proposed extension to encompass `intake-cmip`: we will need to understand the `collection_type` to know which method to call at the `CesmMetadataStoreCatalog` level. 

Let me know what you think. I am happy to accept this PR and address this in a future iteration.

File: intake_cesm/core.py [hunk 2]
```diff
@@ -43,6 +56,24 @@ def set_collection(self, collection):
         """Set the active collection"""
         self.__init__(collection)
 
+    def build_collections(self):
```
Comment (matt-long) [inline intake_cesm/core.py]: Would an alternate implementation be to make this a class method? In that case the API might be something like
```python
col = cesm_metadatastore.build_cesm_metadatastore(**kwarg)
```
whilst preserving the original interface to init. 

It's not obvious to me which is preferable.

File: intake_cesm/manage_collections.py (hunk 1 above)
Comment (andersy005) [reply intake_cesm/manage_collections.py]: These are great ideas as we embark on `intake-cmip` and `intake-cesm` consolidation journey. I opened an issue https://github.com/NCAR/intake-cesm/issues/30 for API discussion. Let's iron out the API implementation details for this in that issue. 

For the time being, I am happy with merging this PR the way it is, and addressing the issues you raised in a separate PR.

Comment (matt-long) [reply intake_cesm/manage_collections.py]: Sounds good.

**Pull request merged**